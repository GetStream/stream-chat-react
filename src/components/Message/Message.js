// @ts-check
import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import { ChannelContext } from '../../context';
import MessageSimple from './MessageSimple';
import { checkChannelPropType, checkClientPropType } from '../../utils';
import {
  MESSAGE_ACTIONS,
  getMessageActions,
  areMessagePropsEqual,
} from './utils';
import {
  useMuteHandler,
  useEditHandler,
  useReactionHandler,
  useDeleteHandler,
  useActionHandler,
  useRetryHandler,
  useMentionsHandler,
  useUserHandler,
  useOpenThreadHandler,
  useUserRole,
  useFlagHandler,
} from './hooks';

/**
 * Message - A high level component which implements all the logic required for a message.
 * The actual rendering of the message is delegated via the "Message" property
 *
 * @example ../../docs/Message.md
 * @type { React.FC<import('types').MessageComponentProps> }
 */
const Message = (props) => {
  const {
    addNotification,
    getFlagMessageErrorNotification,
    getFlagMessageSuccessNotification,
    getMuteUserErrorNotification,
    getMuteUserSuccessNotification,
    message,
    channel: propChannel,
    onUserClick: propOnUserClick,
    onUserHover: propOnUserHover,
    openThread: propOpenThread,
    retrySendMessage: propRetrySendMessage,
    onMentionsClick: propOnMentionsClick,
    onMentionsHover: propOnMentionsHover,
    Message: MessageUIComponent = MessageSimple,
    messageActions = Object.keys(MESSAGE_ACTIONS),
    formatDate,
    groupStyles = [],
  } = props;
  const { channel: contextChannel } = useContext(ChannelContext);
  const channel = propChannel || contextChannel;
  const channelConfig = channel?.getConfig && channel.getConfig();
  const { editing, setEdit, clearEdit } = useEditHandler();
  const handleDelete = useDeleteHandler(message);
  const handleReaction = useReactionHandler(message);
  const handleAction = useActionHandler(message);
  const handleRetry = useRetryHandler(propRetrySendMessage);
  const handleOpenThread = useOpenThreadHandler(message, propOpenThread);
  const handleFlag = useFlagHandler(message, {
    notify: addNotification,
    getSuccessNotification: getFlagMessageSuccessNotification,
    getErrorNotification: getFlagMessageErrorNotification,
  });
  const handleMute = useMuteHandler(message, {
    notify: addNotification,
    getSuccessNotification: getMuteUserSuccessNotification,
    getErrorNotification: getMuteUserErrorNotification,
  });
  const { onMentionsClick, onMentionsHover } = useMentionsHandler(message, {
    onMentionsClick: propOnMentionsClick,
    onMentionsHover: propOnMentionsHover,
  });
  const { onUserClick, onUserHover } = useUserHandler(message, {
    onUserClickHandler: propOnUserClick,
    onUserHoverHandler: propOnUserHover,
  });
  const { isMyMessage, isAdmin, isModerator, isOwner } = useUserRole(message);
  const canReact = true;
  const canReply = true;
  const canEdit = isMyMessage || isModerator || isOwner || isAdmin;
  const canDelete = canEdit;
  const messageActionsHandler = useCallback(() => {
    if (!message || !messageActions) {
      return [];
    }

    return getMessageActions(messageActions, {
      canDelete,
      canEdit,
      canReply,
      canReact,
      canFlag: !isMyMessage,
      canMute: !isMyMessage && !!channelConfig?.mutes,
    });
  }, [
    channelConfig,
    message,
    messageActions,
    canDelete,
    canEdit,
    canReply,
    canReact,
    isMyMessage,
  ]);

  const actionsEnabled =
    message && message.type === 'regular' && message.status === 'received';

  return (
    MessageUIComponent && (
      <MessageUIComponent
        {...props}
        editing={editing}
        formatDate={formatDate}
        clearEditingState={clearEdit}
        setEditingState={setEdit}
        groupStyles={groupStyles}
        actionsEnabled={actionsEnabled}
        Message={MessageUIComponent}
        handleReaction={handleReaction}
        getMessageActions={messageActionsHandler}
        handleFlag={handleFlag}
        handleMute={handleMute}
        handleAction={handleAction}
        handleDelete={handleDelete}
        handleEdit={setEdit}
        handleRetry={handleRetry}
        handleOpenThread={handleOpenThread}
        isMyMessage={() => isMyMessage}
        channelConfig={channelConfig}
        onMentionsClickMessage={onMentionsClick}
        onMentionsHoverMessage={onMentionsHover}
        onUserClick={onUserClick}
        onUserHover={onUserHover}
      />
    )
  );
};

Message.propTypes = {
  /** The message object */
  message: /** @type {PropTypes.Validator<import('stream-chat').MessageResponse>} */ (PropTypes.shape(
    {
      text: PropTypes.string.isRequired,
      html: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      reaction_counts: PropTypes.objectOf(PropTypes.number.isRequired),
      reaction_scores: PropTypes.objectOf(PropTypes.number.isRequired),
      created_at: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
      updated_at: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    },
  ).isRequired),
  /** The client connection object for connecting to Stream */
  client: /** @type {PropTypes.Validator<import('types').StreamChatReactClient>} */ (PropTypes.objectOf(
    checkClientPropType,
  ).isRequired),
  /** The current channel this message is displayed in */
  channel: /** @type {PropTypes.Validator<ReturnType<import('types').StreamChatReactClient['channel']>>} */ (PropTypes.objectOf(
    checkChannelPropType,
  ).isRequired),
  /** A list of users that have read this message */
  readBy: PropTypes.array,
  /** groupStyles, a list of styles to apply to this message. ie. top, bottom, single etc */
  groupStyles: PropTypes.array,

  /** Override the default formatting of the date. This is a function that has access to the original date object. Returns a string or Node  */
  formatDate: PropTypes.func,
  /**
   * Message UI component to display a message in message list.
   * Available from [channel context](https://getstream.github.io/stream-chat-react/#channelcontext)
   * */
  Message: /** @type {PropTypes.Validator<React.ElementType<import('types').MessageUIComponentProps>>} */ (PropTypes.elementType),
  /**
   * The component that will be rendered if the message has been deleted.
   * All props are passed into this component.
   */
  MessageDeleted: /** @type {PropTypes.Validator<React.ElementType<import('types').MessageDeletedProps>>} */ (PropTypes.elementType),

  /**
   * A component to display the selector that allows a user to react to a certain message.
   */
  ReactionSelector: /** @type {PropTypes.Validator<React.ElementType<import('types').ReactionSelectorProps>>} */ (PropTypes.elementType),
  /**
   * A component to display the a message list of reactions.
   */
  ReactionsList: /** @type {PropTypes.Validator<React.ElementType<import('types').ReactionsListProps>>} */ (PropTypes.elementType),
  /**
   * Attachment UI component to display attachment in individual message.
   * Available from [channel context](https://getstream.github.io/stream-chat-react/#channelcontext)
   * */
  Attachment: /** @type {PropTypes.Validator<React.ElementType<import('types').WrapperAttachmentUIComponentProps>>} */ (PropTypes.elementType),
  /** render HTML instead of markdown. Posting HTML is only allowed server-side */
  unsafeHTML: PropTypes.bool,
  /**
   * Array of allowed actions on message. e.g. ['edit', 'delete', 'mute', 'flag']
   * If all the actions need to be disabled, empty array or false should be provided as value of prop.
   * */
  messageActions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  /**
   * Function that returns message/text as string to be shown as notification, when request for flagging a message is successful
   *
   * This function should accept following params:
   *
   * @param message A [message object](https://getstream.io/chat/docs/#message_format) which is flagged.
   *
   * */
  getFlagMessageSuccessNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for flagging a message runs into error
   *
   * This function should accept following params:
   *
   * @param message A [message object](https://getstream.io/chat/docs/#message_format) which is flagged.
   *
   * */
  getFlagMessageErrorNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for muting a user is successful
   *
   * This function should accept following params:
   *
   * @param user A user object which is being muted
   *
   * */
  getMuteUserSuccessNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for muting a user runs into error
   *
   * This function should accept following params:
   *
   * @param user A user object which is being muted
   *
   * */
  getMuteUserErrorNotification: PropTypes.func,
  /** Latest message id on current channel */
  lastReceivedId: PropTypes.string,
  /** DOMRect object for parent MessageList component */
  messageListRect: /** @type {PropTypes.Validator<DOMRect>} */ (PropTypes.object),
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  members: /** @type {PropTypes.Validator<import('seamless-immutable').ImmutableObject<{[user_id: string]: import('stream-chat').ChannelMemberResponse<import('types').StreamChatReactUserType>}> | null | undefined>} */ (PropTypes.object),
  /**
   * Function to add custom notification on messagelist
   *
   * @param text Notification text to display
   * @param type Type of notification. 'success' | 'error'
   * */
  addNotification: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  updateMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  removeMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  retrySendMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  onMentionsClick: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  onMentionsHover: PropTypes.func,
  /**
   * The handler for click event on the user that posted the message
   *
   * @param event Dom click event which triggered handler.
   * @param user the User object for the corresponding user.
   */
  onUserClick: PropTypes.func,
  /**
   * The handler for hover events on the user that posted the message
   *
   * @param event Dom hover event which triggered handler.
   * @param user the User object for the corresponding user.
   */
  onUserHover: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  openThread: PropTypes.func,
  /**
   * Additional props for underlying MessageInput component.
   * Available props - https://getstream.github.io/stream-chat-react/#messageinput
   * */
  additionalMessageInputProps: PropTypes.object,
};

Message.defaultProps = {
  Message: MessageSimple,
  readBy: [],
  groupStyles: [],
  messageActions: Object.keys(MESSAGE_ACTIONS),
};

export default React.memo(Message, areMessagePropsEqual);
