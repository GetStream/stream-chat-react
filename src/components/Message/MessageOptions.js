// @ts-check
import React, { useContext } from 'react';
import { useUserRole, useOpenThreadHandler } from './hooks';
import { ChannelContext } from '../../context';
import { MessageActions } from '../MessageActions';
import { MESSAGE_ACTIONS } from './utils';
import { ThreadIcon, ReactionIcon } from './icons';

/**
 * @type { React.FC<import('types').MessageOptionsProps> }
 */
const MessageOptionsComponent = (props) => {
  const {
    displayActions = true,
    displayLeft = true,
    displayReplies = true,
    handleOpenThread: propHandleOpenThread,
    initialMessage,
    message,
    messageWrapperRef,
    onReactionListClick,
    theme = 'simple',
    threadList,
  } = props;
  const { isMyMessage } = useUserRole(message);
  const handleOpenThread = useOpenThreadHandler(message);
  /**
   * @type {import('types').ChannelContextValue}
   */
  const { channel } = useContext(ChannelContext);
  const channelConfig = channel?.getConfig();
  const messageActions = props.getMessageActions();
  const shouldShowReactions =
    messageActions.indexOf(MESSAGE_ACTIONS.react) > -1 &&
    channelConfig &&
    channelConfig.reactions;

  const shouldShowReplies =
    messageActions.indexOf(MESSAGE_ACTIONS.reply) > -1 &&
    displayReplies &&
    !threadList &&
    channelConfig &&
    channelConfig.replies;

  if (
    !message ||
    message.type === 'error' ||
    message.type === 'system' ||
    message.type === 'ephemeral' ||
    message.status === 'failed' ||
    message.status === 'sending' ||
    initialMessage
  ) {
    return null;
  }
  if (isMyMessage && displayLeft) {
    return (
      <div
        data-testid="message-options-left"
        className={`str-chat__message-${theme}__actions`}
      >
        {<MessageActions {...props} messageWrapperRef={messageWrapperRef} />}
        {shouldShowReplies && (
          <div
            data-testid="thread-action"
            onClick={propHandleOpenThread || handleOpenThread}
            className={`str-chat__message-${theme}__actions__action str-chat__message-${theme}__actions__action--thread`}
          >
            <ThreadIcon />
          </div>
        )}
        {shouldShowReactions && (
          <div
            data-testid="message-reaction-action"
            className={`str-chat__message-${theme}__actions__action str-chat__message-${theme}__actions__action--reactions`}
            onClick={onReactionListClick}
          >
            <ReactionIcon />
          </div>
        )}
      </div>
    );
  }
  return (
    <div
      data-testid="message-options"
      className={`str-chat__message-${theme}__actions`}
    >
      {shouldShowReactions && (
        <div
          data-testid="message-reaction-action"
          className={`str-chat__message-${theme}__actions__action str-chat__message-${theme}__actions__action--reactions`}
          onClick={onReactionListClick}
        >
          <ReactionIcon />
        </div>
      )}
      {shouldShowReplies && (
        <div
          onClick={propHandleOpenThread || handleOpenThread}
          data-testid="thread-action"
          className={`str-chat__message-${theme}__actions__action str-chat__message-${theme}__actions__action--thread`}
        >
          <ThreadIcon />
        </div>
      )}
      {displayActions && (
        <MessageActions {...props} messageWrapperRef={messageWrapperRef} />
      )}
    </div>
  );
};

export default React.memo(MessageOptionsComponent);
