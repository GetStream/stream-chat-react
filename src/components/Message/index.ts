export * from './FixedHeightMessage';
export * from './hooks';
export * from './icons';
export * from './Message';
export * from './MessageDeleted';
export * from './MessageOptions';
export * from './MessageRepliesCountButton';
export * from './MessageSimple';
export * from './MessageStatus';
export * from './MessageText';
export * from './MessageTimestamp';
export * from './QuotedMessage';
export * from './types';
export * from './utils';
