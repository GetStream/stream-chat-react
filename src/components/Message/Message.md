```js
import { Message } from './Message';
import { MessageSimple } from './MessageSimple';
import { Channel } from '../../components/Channel/Channel'
import { Chat } from '../../components/Chat/Chat'
import { MessageContextValue, MessageProvider } from '../../context/MessageContext';

const data = require('../../docs/data');
<div className="str-chat" style={{ height: 'unset' }}>
<Chat client={data.client}>
  <Channel channel={data.channel} Message={MessageSimple}>
    <Message
      message={data.message}  
      readBy={[]}
      groupStyles={['top']}
      editing={false}
      mutes={[]}
      {...data.channelContext}
      {...data.translationContext}
    />
  </Channel>
</Chat>
</div>;
```