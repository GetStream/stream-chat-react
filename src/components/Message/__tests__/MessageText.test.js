/* eslint-disable jest-dom/prefer-to-have-class */
import '@testing-library/jest-dom';
import { act, fireEvent, render, screen, waitFor } from '@testing-library/react';
import EmojiComponentMock from 'emoji-mart/dist-modern/components/emoji/nimble-emoji';
import { toHaveNoViolations } from 'jest-axe';
import React from 'react';
import testRenderer from 'react-test-renderer';
import { axe } from '../../../../axe-helper';

import {
  ChannelActionProvider,
  ChannelStateProvider,
  ChatProvider,
  ComponentProvider,
  EmojiProvider,
  TranslationProvider,
} from '../../../context';
import {
  emojiDataMock,
  generateChannel,
  generateMessage,
  generateReaction,
  generateUser,
  getTestClientWithUser,
} from '../../../mock-builders';

import { Attachment } from '../../Attachment';
import { Message } from '../Message';
import { MessageOptions as MessageOptionsMock } from '../MessageOptions';
import { MessageSimple } from '../MessageSimple';
import { MessageText } from '../MessageText';

expect.extend(toHaveNoViolations);

jest.mock('../MessageOptions', () => ({
  MessageOptions: jest.fn(() => <div />),
}));

const alice = generateUser({ name: 'alice' });
const bob = generateUser({ name: 'bob' });
const onMentionsHoverMock = jest.fn();
const onMentionsClickMock = jest.fn();
const defaultProps = {
  initialMessage: false,
  message: generateMessage(),
  messageWrapperRef: { current: document.createElement('div') },
  onReactionListClick: () => {},
  threadList: false,
};

function generateAliceMessage(messageOptions) {
  return generateMessage({
    user: alice,
    ...messageOptions,
  });
}

async function renderMessageText(customProps, channelConfigOverrides = {}, renderer = render) {
  const client = await getTestClientWithUser(alice);
  const channel = generateChannel({
    getConfig: () => ({ reactions: true, ...channelConfigOverrides }),
    state: { membership: {} },
  });
  const channelCapabilities = { 'send-reaction': true };
  const channelConfig = channel.getConfig();
  const customDateTimeParser = jest.fn(() => ({ format: jest.fn() }));

  return renderer(
    <ChatProvider value={{ client }}>
      <ChannelStateProvider value={{ channel, channelCapabilities, channelConfig }}>
        <ChannelActionProvider
          value={{ onMentionsClick: onMentionsClickMock, onMentionsHover: onMentionsHoverMock }}
        >
          <TranslationProvider
            value={{
              t: (key) => key,
              tDateTimeParser: customDateTimeParser,
              userLanguage: 'en',
            }}
          >
            <ComponentProvider
              value={{
                Attachment,
                Emoji: EmojiComponentMock,
                // eslint-disable-next-line react/display-name
                Message: () => <MessageSimple channelConfig={channelConfig} />,
              }}
            >
              <EmojiProvider value={{ emojiConfig: emojiDataMock }}>
                <Message {...defaultProps} {...customProps}>
                  <MessageText {...defaultProps} {...customProps} />
                </Message>
              </EmojiProvider>
            </ComponentProvider>
          </TranslationProvider>
        </ChannelActionProvider>
      </ChannelStateProvider>
    </ChatProvider>,
  );
}

const messageTextTestId = 'message-text-inner-wrapper';
const reactionSelectorTestId = 'reaction-selector';

describe('<MessageText />', () => {
  beforeEach(jest.clearAllMocks);
  it('should not render anything if message is not set', async () => {
    const { container, queryByTestId } = await renderMessageText({ message: {} });
    expect(queryByTestId(messageTextTestId)).not.toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should not render anything if message text is not set', async () => {
    const { container, queryByTestId } = await renderMessageText({ message: {} });
    expect(queryByTestId(messageTextTestId)).not.toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should set attachments css class modifier when message has text and is focused', async () => {
    const attachment = {
      image_url: 'http://image.jpg',
      type: 'image',
    };
    const message = generateAliceMessage({
      attachments: [attachment, attachment, attachment],
    });
    const { container, getByTestId } = await renderMessageText({ message });
    expect(getByTestId(messageTextTestId).className).toContain('--has-attachment');
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should set emoji css class when message has text that is only emojis', async () => {
    const message = generateAliceMessage({ text: '🤖🤖🤖🤖' });
    const { container, getByTestId } = await renderMessageText({ message });
    expect(getByTestId(messageTextTestId).className).toContain('--is-emoji');
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should handle message mention mouse hover event', async () => {
    const message = generateAliceMessage({ mentioned_users: [bob] });
    const { container, getByTestId } = await renderMessageText({
      message,
      onMentionsHoverMessage: onMentionsHoverMock,
    });
    expect(onMentionsHoverMock).not.toHaveBeenCalled();
    fireEvent.mouseOver(getByTestId(messageTextTestId));
    expect(onMentionsHoverMock).toHaveBeenCalledTimes(1);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should handle message mention mouse click event', async () => {
    const message = generateAliceMessage({ mentioned_users: [bob] });
    const { container, getByTestId } = await renderMessageText({
      message,
      onMentionsClickMessage: onMentionsClickMock,
    });
    expect(onMentionsClickMock).not.toHaveBeenCalled();
    fireEvent.click(getByTestId(messageTextTestId));
    expect(onMentionsClickMock).toHaveBeenCalledTimes(1);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should inform that message was not sent when message is has type "error"', async () => {
    const message = generateAliceMessage({ type: 'error' });
    const { container, getByText } = await renderMessageText({ message });
    expect(getByText('Error · Unsent')).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should inform that retry is possible when message has status "failed"', async () => {
    const message = generateAliceMessage({ status: 'failed' });
    const { container, getByText } = await renderMessageText({ message });
    expect(getByText('Message Failed · Click to try again')).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('render message html when unsafe html property is enabled', async () => {
    const message = generateAliceMessage({
      html: '<span data-testid="custom-html" />',
    });
    const { container, getByTestId } = await renderMessageText({
      message,
      unsafeHTML: true,
    });
    expect(getByTestId('custom-html')).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('render message text', async () => {
    const text = 'Hello, world!';
    const message = generateAliceMessage({ text });
    const { container, getByText } = await renderMessageText({ message });
    expect(getByText(text)).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should display text in users set language', async () => {
    const text = 'bonjour';
    const message = generateAliceMessage({
      i18n: { en_text: 'hello', fr_text: 'bonjour', language: 'fr' },
      text,
    });

    const { container, getByText } = await renderMessageText({ message });

    expect(getByText('hello')).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should show reaction list if message has reactions and detailed reactions are not displayed', async () => {
    const bobReaction = generateReaction({ user: bob });
    const message = generateAliceMessage({
      latest_reactions: [bobReaction],
    });

    let container;
    await act(async () => {
      const result = await renderMessageText({ message });
      container = result.container;
    });

    await waitFor(() => expect(screen.getByTestId('reaction-list')).toBeInTheDocument());
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should not show reaction list if disabled in channelConfig', async () => {
    const bobReaction = generateReaction({ user: bob });
    const message = generateAliceMessage({
      latest_reactions: [bobReaction],
    });
    const { container, queryByTestId } = await renderMessageText({ message }, { reactions: false });

    expect(queryByTestId('reaction-list')).not.toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should show reaction selector when message has reaction and reaction list is clicked', async () => {
    const bobReaction = generateReaction({ user: bob });
    const message = generateAliceMessage({
      latest_reactions: [bobReaction],
    });
    const { container, getByTestId, queryByTestId } = await renderMessageText({ message });
    expect(queryByTestId(reactionSelectorTestId)).not.toBeInTheDocument();
    await act(() => {
      fireEvent.click(getByTestId('reaction-list'));
    });

    expect(getByTestId(reactionSelectorTestId)).toBeInTheDocument();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render message options', async () => {
    const { container } = await renderMessageText();
    expect(MessageOptionsMock).toHaveBeenCalledTimes(1);
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render message options with custom props when those are set', async () => {
    const displayLeft = false;
    const { container } = await renderMessageText({
      customOptionProps: {
        displayLeft,
      },
    });
    // eslint-disable-next-line jest/prefer-called-with
    expect(MessageOptionsMock).toHaveBeenCalled();
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('should render with a custom wrapper class when one is set', async () => {
    const customWrapperClass = 'custom-wrapper';
    const message = generateMessage({ text: 'hello world' });
    const tree = await renderMessageText({ customWrapperClass, message }, {}, testRenderer.create);
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
      >
        <div
          className="str-chat__message-inner"
          data-testid="message-inner"
        >
          <div />
          <div
            className="str-chat__message-reactions-host"
          />
          <div
            className="str-chat__message-bubble"
          >
            <div
              className="str-chat__message-text"
              tabIndex={0}
            >
              <div
                className="str-chat__message-text-inner str-chat__message-simple-text-inner"
                data-testid="message-text-inner-wrapper"
                onClick={[Function]}
                onMouseOver={[Function]}
              >
                <div>
                  <p>
                    hello world
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    `);
  });

  it('should render with a custom inner class when one is set', async () => {
    const customInnerClass = 'custom-inner';
    const message = generateMessage({ text: 'hi mate' });
    const tree = await renderMessageText({ customInnerClass, message }, {}, testRenderer.create);
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
      >
        <div
          className="str-chat__message-inner"
          data-testid="message-inner"
        >
          <div />
          <div
            className="str-chat__message-reactions-host"
          />
          <div
            className="str-chat__message-bubble"
          >
            <div
              className="str-chat__message-text"
              tabIndex={0}
            >
              <div
                className="str-chat__message-text-inner str-chat__message-simple-text-inner"
                data-testid="message-text-inner-wrapper"
                onClick={[Function]}
                onMouseOver={[Function]}
              >
                <div>
                  <p>
                    hi mate
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    `);
  });

  it('should render with custom theme identifier in generated css classes when theme is set', async () => {
    const message = generateMessage({ text: 'whatup?!' });
    const tree = await renderMessageText({ message, theme: 'custom' }, {}, testRenderer.create);
    expect(tree.toJSON()).toMatchInlineSnapshot(`
      <div
        className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
      >
        <div
          className="str-chat__message-inner"
          data-testid="message-inner"
        >
          <div />
          <div
            className="str-chat__message-reactions-host"
          />
          <div
            className="str-chat__message-bubble"
          >
            <div
              className="str-chat__message-text"
              tabIndex={0}
            >
              <div
                className="str-chat__message-text-inner str-chat__message-simple-text-inner"
                data-testid="message-text-inner-wrapper"
                onClick={[Function]}
                onMouseOver={[Function]}
              >
                <div>
                  <p>
                    whatup?!
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    `);
  });
});
