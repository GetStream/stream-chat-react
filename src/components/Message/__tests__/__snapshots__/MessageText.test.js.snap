// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageText /> should render with a custom inner class when one is set 1`] = `
<div
  className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
>
  <div
    className="str-chat__message-inner"
    data-testid="message-inner"
  >
    <div />
    <div
      className="str-chat__message-reactions-host"
    />
    <div
      className="str-chat__message-bubble"
    >
      <div
        className="str-chat__message-text"
        tabIndex={0}
      >
        <div
          className="str-chat__message-text-inner str-chat__message-simple-text-inner"
          data-testid="message-text-inner-wrapper"
          onClick={[Function]}
          onMouseOver={[Function]}
        >
          <div>
            <p>
              hi mate
            </p>
          </div>
        </div>
      </div>
      <div
        className="str-chat__message-error-icon"
      >
        <svg
          data-testid="error"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2Z"
            fill="black"
            id="background"
          />
          <path
            d="M13 17H11V15H13V17ZM13 13H11V7H13V13Z"
            fill="white"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="str-chat__message-metadata"
  />
</div>
`;

exports[`<MessageText /> should render with a custom wrapper class when one is set 1`] = `
<div
  className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
>
  <div
    className="str-chat__message-inner"
    data-testid="message-inner"
  >
    <div />
    <div
      className="str-chat__message-reactions-host"
    />
    <div
      className="str-chat__message-bubble"
    >
      <div
        className="str-chat__message-text"
        tabIndex={0}
      >
        <div
          className="str-chat__message-text-inner str-chat__message-simple-text-inner"
          data-testid="message-text-inner-wrapper"
          onClick={[Function]}
          onMouseOver={[Function]}
        >
          <div>
            <p>
              hello world
            </p>
          </div>
        </div>
      </div>
      <div
        className="str-chat__message-error-icon"
      >
        <svg
          data-testid="error"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2Z"
            fill="black"
            id="background"
          />
          <path
            d="M13 17H11V15H13V17ZM13 13H11V7H13V13Z"
            fill="white"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="str-chat__message-metadata"
  />
</div>
`;

exports[`<MessageText /> should render with custom theme identifier in generated css classes when theme is set 1`] = `
<div
  className="str-chat__message str-chat__message-simple str-chat__message--regular str-chat__message--received str-chat__message--other str-chat__message--has-text"
>
  <div
    className="str-chat__message-inner"
    data-testid="message-inner"
  >
    <div />
    <div
      className="str-chat__message-reactions-host"
    />
    <div
      className="str-chat__message-bubble"
    >
      <div
        className="str-chat__message-text"
        tabIndex={0}
      >
        <div
          className="str-chat__message-text-inner str-chat__message-simple-text-inner"
          data-testid="message-text-inner-wrapper"
          onClick={[Function]}
          onMouseOver={[Function]}
        >
          <div>
            <p>
              whatup?!
            </p>
          </div>
        </div>
      </div>
      <div
        className="str-chat__message-error-icon"
      >
        <svg
          data-testid="error"
          fill="none"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12C2 17.52 6.48 22 12 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 12 2Z"
            fill="black"
            id="background"
          />
          <path
            d="M13 17H11V15H13V17ZM13 13H11V7H13V13Z"
            fill="white"
          />
        </svg>
      </div>
    </div>
  </div>
  <div
    className="str-chat__message-metadata"
  />
</div>
`;
